### DISCUSSION TITLE
improve robot access token file format

### DISCUSSION THREAD

**Discussion by toxeus** [goharbor/harbor#14390] [Ideas]
After creating a new robot account it's possible to download the generated token as a file. Unfortunately, the JSON formatting doesn't allow the file to be used seamlessly with

```sh
docker login --password-stdin < 'robot$something.json'
```
.
 
A clear improvement would be if the file would contain the access token exclusively. 
The proposed change is trivial to implement.

**Comment by rrgadeev** [comment]
cat robot\$project+test.json | jq .secret

**Comment by rrgadeev** [comment]
docker login harbor.domainname.local --username $(jq -r .name robot\$project+test.json) --password $(jq -r .secret  robot\$project+test.json)

**Comment by toxeus** [comment]
Yes, I know about `jq`. But this tool is not installed by default on most distros. I'm distributing the robot files to third parties and it would be great if it would just work out-of-the-box instead of me explaining to them how to install and use `jq` correctly.

Also, your code snippets are wrong: you didn't escape `$` so everything following it will be interpreted as an env var name and expanded to an empty string since those env vars are most likely undefined. That nicely supports my argument above.

Using `$` in the user and filename is another design choice (orthogonal to this one) that impacts UX for no reason that is apparent to me. But I'm hesitant to open an issue about that one since fixing it impacts backwards compatibility for existing user names.

**Comment by rrgadeev** [comment]
> Yes, I know about `jq`. But this tool is not installed by default on most distros. I'm distributing the robot files to third parties and it would be great if it would just work out-of-the-box instead of me explaining to them how to install and use `jq` correctly.
> 
> Also, your code snippets are wrong: you didn't escape `$` so everything following it will be interpreted as an env var name and expanded to an empty string since those env vars are most likely undefined. That nicely supports my argument above.
> 
> Using `$` in the user and filename is another design choice (orthogonal to this one) that impacts UX for no reason that is apparent to me. But I'm hesitant to open an issue about that one since fixing it impacts backwards compatibility for existing user names.
it works, $() - https://stackoverflow.com/questions/27472540/difference-between-and-in-bash
I think if people can use docker, they can also use jq

**Comment by toxeus** [comment]
Thx for re-confirming my argument :blush: 
![image](https://user-images.githubusercontent.com/7702854/109793019-6aec6700-7c14-11eb-9845-d04abd510a29.png)

**Comment by rrgadeev** [comment]
![image](https://user-images.githubusercontent.com/67308688/109793580-6a20f880-7c47-11eb-83d8-7625fea9a4f5.png)

**Comment by rrgadeev** [comment]
you are right, there itself was escaped in the process of pressing tab, I forgot ``

**Comment by rrgadeev** [comment]
`docker login harbor.domainname.local --username $(jq -r .name robot\$project+test.json) --password $(jq -r .secret robot\$project+test.json) `

**Comment by toxeus** [comment]
I guess from the discussion we just had now everyone can understand how guiding of third-parties can look like. Yes, we eventually manage to make it work. But the more such cases I have the more time-consuming it is.

If there's no good reason for the current file format then I'd change it to enable a better out-of-the-box experience. I'm surprised that this is controversial since it should be in the interest of the project to improve UX.

**Comment by phin1x** [comment]
I would disagree that changing the file format would improve UX. Exporting the robot data in json you have the username and the token in one file (better for distributing), the credentials are better to parse (like the docker login example above) and with jq they are easy to read. nowadays jq is part of the standard toolchain.

Whether it is better to export the information as plain text or as json, depends then nevertheless strongly on the use case and the personal opinion. i prefer the json format.

**Comment by ninjadq** [comment]
Convert it discussion as per there are a lot of discuss
